### PR TITLE
Dispvm 4.0 updates

### PR DESCRIPTION
**Pull request by awokd**
Sourced primarily from https://github.com/QubesOS/qubes-issues/issues/2253 and https://groups.google.com/forum/?_escaped_fragment_=topic/qubes-devel/UGh8NDdkrXo#!topic/qubes-devel/UGh8NDdkrXo

### CODE REVIEW

File: common-tasks/dispvm.md [hunk 1]
```diff
@@ -20,7 +20,31 @@ Once a DispVM has been created it will appear in Qubes VM Manager with the name
...
+Thus if an AppVM uses sys-net as its NetVM, but the default system DispVM uses sys-whonix,
+any DispVM launched from this AppVM will have sys-whonix as its NetVM.
+The default system wide DispVM template can be changed with `qubes-prefs default_dispvm`.
+You can change this behaviour for individual VMs: in the Application Menu, open Qube Settings
+for the VM in question and go to the "Advanced" tab. 
+Here you can edit the "Default DispVM" setting to specify which DispVM template will be used to launch DispVMs from that VM.
+Disposable VMs will temporarily appear with the name `disp####`. 
+
+A Disposable VM launched from the Start Menu inherits the NetVM and firewall settings of the [DVM Template](https://www.qubes-os.org/doc/glossary/#dvm-template) from which it is built. 
+By default the DVM template is called `fedora-XX-dvm` (where `XX` is the Fedora version of the default TemplateVM).
```
Comment (marmarek) [inline common-tasks/dispvm.md]: Those two sentences looks duplicated with the first paragraph.

File: common-tasks/dispvm.md [hunk 2]
```diff
@@ -20,7 +20,31 @@ Once a DispVM has been created it will appear in Qubes VM Manager with the name
...
+any DispVM launched from this AppVM will have sys-whonix as its NetVM.
+The default system wide DispVM template can be changed with `qubes-prefs default_dispvm`.
+You can change this behaviour for individual VMs: in the Application Menu, open Qube Settings
+for the VM in question and go to the "Advanced" tab. 
+Here you can edit the "Default DispVM" setting to specify which DispVM template will be used to launch DispVMs from that VM.
+Disposable VMs will temporarily appear with the name `disp####`. 
+
+A Disposable VM launched from the Start Menu inherits the NetVM and firewall settings of the [DVM Template](https://www.qubes-os.org/doc/glossary/#dvm-template) from which it is built. 
+By default the DVM template is called `fedora-XX-dvm` (where `XX` is the Fedora version of the default TemplateVM). 
+As an "internal" VM it is hidden in Qubes VM Manager, but can be shown by selecting "Show/Hide internal VMs".
```
Comment (marmarek) [inline common-tasks/dispvm.md]: fedora-26-dvm is not marked as "internal" by default in 4.0. Also, Qube Manger in 4.0 no longer hide "internal" VMs.

File: common-tasks/dispvm.md (hunk 1 above)
Comment (awokd) [reply common-tasks/dispvm.md]: In the first paragraph of the R4.0 section, I was trying to introduce the R4.0 concept of multiple disposable templates vs. R3.2's single. The paragraph you cited is to spell out the differences between R4.0 and R3.2. I'll take out the (e.g.) from the first para to lessen duplication, but is there a better way the first para could be worded?

File: common-tasks/dispvm.md [hunk 3]
```diff
@@ -20,7 +20,30 @@ Once a DispVM has been created it will appear in Qubes VM Manager with the name
...
+
+R4.0 introduces the concept of multiple disposable VM templates (R3.2 was limited to one).
+This allows for the creation of multiple differently configured disposable VMs that can be accessed from 
+the Applications menu. Even more types of DispVMs can be created on-the-fly on a per AppVM basis.
+As you can see, this is a very flexible and powerful system for managing your Disposable VMs.
+
+NetVM and firewall rules for Disposable VMs can be set as they can for a normal VM. 
+By default a DispVM will inherit the NetVM and firewall settings of the DispVM Template from which it is built. 
+Thus if an AppVM uses sys-net as its NetVM, but the default system DispVM uses sys-whonix,
+any DispVM launched from this AppVM will have sys-whonix as its NetVM.
```
Comment (andrewdavidwong) [inline common-tasks/dispvm.md]: Aren't these two sentences contradictory?

Comment (andrewdavidwong) [reply common-tasks/dispvm.md]: Also, I think we need to first explain that an AppVM can be a DVM Template.

Comment (awokd) [reply common-tasks/dispvm.md]: Not contradictory- it's saying that under 4.0, launching a DispVM from an AppVM will result in it using the DispVM's network/firewall settings (which default to the DVM template from which it was built). Compare & contrast with the same section a little further down under 3.2 which has different behaviour. It's definitely hard to follow, but I'm not sure how else to phrase it...

Comment (awokd) [reply common-tasks/dispvm.md]: I could add "in other words" and include that same sentence from my comment, if it helps understand what's going on?

File: common-tasks/dispvm.md [hunk 4]
```diff
@@ -20,7 +20,30 @@ Once a DispVM has been created it will appear in Qubes VM Manager with the name
 
 See [this article](https://blog.invisiblethings.org/2010/06/01/disposable-vms.html) for more on why one would want to use a Disposable VM.
 
-Disposable VMs and Networking
+Disposable VMs and Networking (R4.0 and later)
+-----------------------------
+
+
+R4.0 introduces the concept of multiple disposable VM templates (R3.2 was limited to one).
```
Comment (andrewdavidwong) [inline common-tasks/dispvm.md]: This would be a great place to introduce the term "DVM Template" in this document, with a link to its glossary entry.

File: common-tasks/dispvm.md [hunk 5]
```diff
@@ -20,7 +20,30 @@ Once a DispVM has been created it will appear in Qubes VM Manager with the name
...
+This allows for the creation of multiple differently configured disposable VMs that can be accessed from 
+the Applications menu. Even more types of DispVMs can be created on-the-fly on a per AppVM basis.
```
Comment (andrewdavidwong) [inline common-tasks/dispvm.md]: Most readers won't understand what this means unless they already know how the new system works.

File: common-tasks/dispvm.md [hunk 6]
```diff
@@ -20,7 +20,30 @@ Once a DispVM has been created it will appear in Qubes VM Manager with the name
...
+As you can see, this is a very flexible and powerful system for managing your Disposable VMs.
+
+NetVM and firewall rules for Disposable VMs can be set as they can for a normal VM. 
+By default a DispVM will inherit the NetVM and firewall settings of the DispVM Template from which it is built.
```
Comment (andrewdavidwong) [inline common-tasks/dispvm.md]: "DispVM Template" should be changed to "DVM Template" so that we use consistent terminology throughout the documentation.

Comment (andrewdavidwong) [reply common-tasks/dispvm.md]: Also, "from which it is built" might be a bit misleading. I think it's more common to say, "on which it is based."

File: common-tasks/dispvm.md (hunk 3 above)
Comment (andrewdavidwong) [reply common-tasks/dispvm.md]: We need to clearly explain the difference between the "based on" and "launched from" relationships, and we need provide examples for the other main cases.

In your example, I launch a DispVM from an AppVM using `sys-net`, but the DispVM doesn't use `sys-net`. Instead, it uses `sys-whonix`, because it's based on my default DVM Template, which uses `sys-whonix`. I can imagine a user having several questions about this:
 - Will the same pattern of behavior occur if I launch a DispVM from `anon-whonix` while my default DVM Template uses `sys-net`? That would be incredibly dangerous.
 - Why was my DispVM based on the default DVM Template rather than the AppVM from which I launched it?
 - How do I base the DispVM on the AppVM instead?

Some of these questions are answered below, but in a way that at least partially contradicts what you wrote above (e.g., because it depends on the settings explained below).

File: common-tasks/dispvm.md [hunk 7]
```diff
@@ -20,7 +20,30 @@ Once a DispVM has been created it will appear in Qubes VM Manager with the name
...
+This allows for the creation of multiple differently configured disposable VMs that can be accessed from 
+the Applications menu. Even more types of DispVMs can be created on-the-fly on a per AppVM basis.
+As you can see, this is a very flexible and powerful system for managing your Disposable VMs.
+
+NetVM and firewall rules for Disposable VMs can be set as they can for a normal VM. 
+By default a DispVM will inherit the NetVM and firewall settings of the DispVM Template from which it is built. 
+Thus if an AppVM uses sys-net as its NetVM, but the default system DispVM uses sys-whonix,
+any DispVM launched from this AppVM will have sys-whonix as its NetVM.
+The default system wide DispVM template can be changed with `qubes-prefs default_dispvm`.
+You can change this behaviour for individual VMs: in the Application Menu, open Qube Settings
```
Comment (andrewdavidwong) [inline common-tasks/dispvm.md]: Grammatically, "this behavior" seems to refer to "the default system wide DispVM template," which is problematic for two reasons:

1. "The default system wide DispVM template" is not a behavior.
2. You can't actually change it in the Qube Settings of individual VMs.

File: common-tasks/dispvm.md [hunk 8]
```diff
@@ -20,7 +20,30 @@ Once a DispVM has been created it will appear in Qubes VM Manager with the name
...
+for the VM in question and go to the "Advanced" tab. 
+Here you can edit the "Default DispVM" setting to specify which DispVM template will be used to launch DispVMs from that VM.
```
Comment (andrewdavidwong) [inline common-tasks/dispvm.md]: Out of curiosity, can you set it to itself (i.e., make it so that all DispVMs launched from an AppVM are based on that AppVM)? If so, it might be worth mentioning this case.

File: common-tasks/dispvm.md [hunk 9]
```diff
@@ -20,7 +20,30 @@ Once a DispVM has been created it will appear in Qubes VM Manager with the name
...
+Disposable VMs will temporarily appear with the name `disp####`.
```
Comment (andrewdavidwong) [inline common-tasks/dispvm.md]: What does this mean? It almost sounds as if they momentarily appear with that strange name right after you edit the Qube Settings, but I don't think that's the case.

If all you mean is that DispVMs are themselves disposable, please don't say it this way.

File: common-tasks/dispvm.md [hunk 10]
```diff
@@ -20,7 +20,30 @@ Once a DispVM has been created it will appear in Qubes VM Manager with the name
...
+Thus if an AppVM uses sys-net as its NetVM, but the default system DispVM uses sys-whonix,
+any DispVM launched from this AppVM will have sys-whonix as its NetVM.
+The default system wide DispVM template can be changed with `qubes-prefs default_dispvm`.
+You can change this behaviour for individual VMs: in the Application Menu, open Qube Settings
+for the VM in question and go to the "Advanced" tab. 
+Here you can edit the "Default DispVM" setting to specify which DispVM template will be used to launch DispVMs from that VM.
+Disposable VMs will temporarily appear with the name `disp####`. 
+
+A Disposable VM launched from the Start Menu inherits the NetVM and firewall settings of the [DVM Template](https://www.qubes-os.org/doc/glossary/#dvm-template) from which it is built. 
+By default the DVM template is called `fedora-XX-dvm` (where `XX` is the Fedora version of the default TemplateVM).
```
Comment (andrewdavidwong) [inline common-tasks/dispvm.md]: Do you mean "the default DVM Template" rather than "by default the DVM Template"? Since there can be more than one DVM Template now, there's no "*the* DVM Template" anymore, though there can be "the *default* DVM Template."

File: common-tasks/dispvm.md [hunk 11]
```diff
@@ -20,7 +20,30 @@ Once a DispVM has been created it will appear in Qubes VM Manager with the name
...
+any DispVM launched from this AppVM will have sys-whonix as its NetVM.
+The default system wide DispVM template can be changed with `qubes-prefs default_dispvm`.
+You can change this behaviour for individual VMs: in the Application Menu, open Qube Settings
+for the VM in question and go to the "Advanced" tab. 
+Here you can edit the "Default DispVM" setting to specify which DispVM template will be used to launch DispVMs from that VM.
+Disposable VMs will temporarily appear with the name `disp####`. 
+
+A Disposable VM launched from the Start Menu inherits the NetVM and firewall settings of the [DVM Template](https://www.qubes-os.org/doc/glossary/#dvm-template) from which it is built. 
+By default the DVM template is called `fedora-XX-dvm` (where `XX` is the Fedora version of the default TemplateVM). 
+Note that changing the "NetVM" setting for the DVM Template *does* affect the NetVM of DispVMs launched from the Start Menu.
```
Comment (andrewdavidwong) [inline common-tasks/dispvm.md]: Before moving on to 3.2, it would be good to say a bit about how having multiple dedicated DVM Templates works.

**Comment by awokd** [comment]
>  Will the same pattern of behavior occur if I launch a DispVM from anon-whonix while my default DVM Template uses sys-net? That would be incredibly dangerous.

The answer to this is "yes", by design (and from a test I just did to confirm.)

>    Why was my DispVM based on the default DVM Template rather than the AppVM from which I launched it?

Because that's R4.0's design? Not sure how you want me to answer this one...

>    How do I base the DispVM on the AppVM instead?

I can answer this one! 

> Some of these questions are answered below, but in a way that at least partially contradicts what you wrote above (e.g., because it depends on the settings explained below).

I see your point on that "behaviour" sentence, but maybe I'm being slow because I'm not seeing any other settings/answers in your comments that contradict what I wrote? I'll go through and attempt to address one by one, maybe it will fix itself.

**Comment by awokd** [comment]
I need to do a fresh install to confirm anon-whonix's default settings, unless either of you know already. Can't recall what its `default_dispvm` defaults to. Tried to address the other points raised.

File: common-tasks/dispvm.md [hunk 12]
```diff
@@ -11,16 +11,56 @@ redirect_from:
...
+----------
+
+Similarly to how AppVMs are based on their underlying [TemplateVM](https://www.qubes-os.org/doc/glossary/#templatevm), DispVMs are based on their underlying [DVM Template](https://www.qubes-os.org/doc/glossary/#dvm-template).
+
+On a fresh installation of Qubes, the DVM template is called `fedora-XX-dvm` (where `XX` is the Fedora version of the default TemplateVM). 
+
+Disposable VMs and Networking (R4.0 and later)
+-----------------------------
+
+R4.0 introduces the concept of multiple disposable VM templates, whereas R3.2 was limited to only one.
```
Comment (andrewdavidwong) [inline common-tasks/dispvm.md]: Why shift from "DVM Templates" to "disposable VM templates" here? That just invites confusion, because now it's not clear whether you're talking about the thing you just introduced, or something slightly different.

**Comment by andrewdavidwong** [comment]
>>  Will the same pattern of behavior occur if I launch a DispVM from anon-whonix while my default DVM Template uses sys-net? That would be incredibly dangerous.
>>
> The answer to this is "yes", by design (and from a test I just did to confirm.)

Ok, that must be an oversight. Created https://github.com/QubesOS/qubes-issues/issues/3561 to address it.

>>    Why was my DispVM based on the default DVM Template rather than the AppVM from which I launched it?
>>
> Because that's R4.0's design? Not sure how you want me to answer this one...

You just did! The key is to explain how the behavior has changed and anticipate things that might be counterintuitive to the reader.

>>    How do I base the DispVM on the AppVM instead?
>>
> I can answer this one! 

:slightly_smiling_face: 

>> Some of these questions are answered below, but in a way that at least partially contradicts what you wrote above (e.g., because it depends on the settings explained below).
>>
> I see your point on that "behaviour" sentence, but maybe I'm being slow because I'm not seeing any other settings/answers in your comments that contradict what I wrote? I'll go through and attempt to address one by one, maybe it will fix itself.

Ok, thanks! Maybe that was the only one. I'll check to see whether I can find any others.

**Comment by andrewdavidwong** [comment]
> I need to do a fresh install to confirm anon-whonix's default settings, unless either of you know already. Can't recall what its default_dispvm defaults to. Tried to address the other points raised.

Ok, pointed out the need for confirmation on the issue.

**Comment by andrewdavidwong** [comment]
Looks like there's a small conflict that needs to be resolved. (Not sure how that happened.)

File: common-tasks/dispvm.md [hunk 13]
```diff
@@ -11,16 +11,56 @@ redirect_from:
...
 See [this article](https://blog.invisiblethings.org/2010/06/01/disposable-vms.html) for more on why one would want to use a Disposable VM.
 
-Disposable VMs and Networking
+
+DVM Templates
+----------
+
+Similarly to how AppVMs are based on their underlying [TemplateVM](https://www.qubes-os.org/doc/glossary/#templatevm), DispVMs are based on their underlying [DVM Template](https://www.qubes-os.org/doc/glossary/#dvm-template).
+
+On a fresh installation of Qubes, the DVM template is called `fedora-XX-dvm` (where `XX` is the Fedora version of the default TemplateVM).
```
Comment (andrewdavidwong) [inline common-tasks/dispvm.md]: How about "the default DVM Template," since there can be more than one? (Also note the capitalization.)

File: common-tasks/dispvm.md [hunk 14]
```diff
@@ -11,16 +11,56 @@ redirect_from:
...
+R4.0 introduces the concept of multiple disposable VM templates, whereas R3.2 was limited to only one.
+You can set any AppVM to have the ability to act as a DVM Template with:
+
+    qvm-prefs <vmname> template_for_dispvms true
+
+The default system wide DVM template can be changed with `qubes-prefs default_dispvm`.
+By combining the two, choosing `Open in Disposable VM` from inside an AppVM will open the document in a DispVM based on the default DVM template you specified.
+
+You can change this behaviour for individual VMs: in the Application Menu, open Qube Settings for the VM in question and go to the "Advanced" tab. 
+Here you can edit the "Default DispVM" setting to specify which DispVM template will be used to launch DispVMs from that VM.
```
Comment (andrewdavidwong) [inline common-tasks/dispvm.md]: "DispVM template" should be "DVM Template."

File: common-tasks/dispvm.md [hunk 15]
```diff
@@ -11,16 +11,56 @@ redirect_from:
...
+This can also be changed from the command line with:
+
+    qvm-prefs <vmname> default_dispvm <dvmtemplatename>
+
+You can even set an AppVM that has also been configured as a DVM template to use itself, so DispVMs launched from within the AppVM/DVM Template would inherit the same settings.
+
+NetVM and firewall rules for DVM templates can be set as they can for a normal VM. 
+By default a DispVM will inherit the NetVM and firewall settings of the DVM Template on which it is based.
+Launching a DispVM from an AppVM will result in it using the DispVM's network/firewall settings (which default to the DVM template on which it is based).
```
Comment (andrewdavidwong) [inline common-tasks/dispvm.md]: Is it really meaningful to say that the DispVM uses the DispVM's network and firewall settings? It doesn't have those settings before it's created. It has to get them from somewhere. If it gets them from the DVM Template, then the DispVM is really using the DVM Template's network and firewall settings, not its own.

File: common-tasks/dispvm.md [hunk 16]
```diff
@@ -92,8 +137,8 @@ However, it is possible to start an arbitrary application in a DispVM directly f
 Customizing Disposable VMs
 --------------------------
 
-You can change the template used to generate the Disposable VM, and change settings used in the Disposable VM savefile. 
-These changes will be reflected in every new Disposable VM. 
+You can change the template used to generate the Disposable VMs, and change settings used in the Disposable VM savefile. 
+These changes will be reflected in every new Disposable VM spawned from that template.
```
Comment (andrewdavidwong) [inline common-tasks/dispvm.md]: Is "spawned from" different from "launched from" and "based on"? "Spawned from" suggests that "that template" (a DVM Template?) is the parent, and the DispVM is the child, which is usually how we describe the relationship between TemplateVMs and TemplateBasedVMs. So, I'm guessing that "spawned from" is synonymous with "based on," and both are distinct from "launched from."

File: common-tasks/dispvm.md (hunk 15 above)
Comment (awokd) [reply common-tasks/dispvm.md]: It's meaningful because they can be changed after the fact. Will add somewhere.

Comment (awokd) [reply common-tasks/dispvm.md]: ... but after testing I'm not sure it takes immediate effect, which would make it useless information.
 What does "NetVM and firewall rules for Disposable VMs can be set as they can for a normal VM." mean in the R3.2 section?

Comment (andrewdavidwong) [reply common-tasks/dispvm.md]: > What does "NetVM and firewall rules for Disposable VMs can be set as they can for a normal VM." mean in the R3.2 section?

I interpret that as meaning that you can chose a NetVM and add or remove firewall rules for a DispVM in the same way you can for a regular TemplateBasedVM.

**Comment by awokd** [comment]
>>    What does "NetVM and firewall rules for Disposable VMs can be set as they can for a normal VM." mean in the R3.2 section?
> I interpret that as meaning that you can chose a NetVM and add or remove firewall rules for a DispVM in the same way you can for a regular TemplateBasedVM.

OK; I removed the equivalent I wrote for R4.0 with my last commit because it doesn't make sense there then. Hopefully that's everything!

**Comment by andrewdavidwong** [comment]
Thanks!
